### ISSUE TITLE
Fatal Error in queue-async

### ISSUE THREAD

**Issue by rafatower** [CartoDB/Windshaft#693]
```
[2019-06-12 04:56:25.096] [FATAL] [default] - Error
    at /home/ubuntu/www/node-windshaft/releases/20190521103931/node_modules/queue-async/build/queue.js:39:41
    at cb (/home/ubuntu/www/node-windshaft/releases/20190521103931/node_modules/windshaft/lib/windshaft/renderers/blend/factory.js:54:28)
    at /home/ubuntu/www/node-windshaft/releases/20190521103931/node_modules/windshaft/lib/windshaft/renderers/renderer_factory.js:89:24
    at mmlBuilder.toXML (/home/ubuntu/www/node-windshaft/releases/20190521103931/node_modules/windshaft/lib/windshaft/renderers/mapnik/factory.js:226:20)
    at ChildProcess.done (/home/ubuntu/www/node-windshaft/releases/20190521103931/node_modules/grainstore/lib/grainstore/mml-builder/mml-builder.js:102:20)
    at Object.onceWrapper (events.js:277:13)
    at ChildProcess.emit (events.js:189:13)
    at emit (internal/child_process.js:820:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
Error
    at /home/ubuntu/www/node-windshaft/releases/20190521103931/node_modules/queue-async/build/queue.js:39:41
    at cb (/home/ubuntu/www/node-windshaft/releases/20190521103931/node_modules/windshaft/lib/windshaft/renderers/blend/factory.js:54:28)
    at /home/ubuntu/www/node-windshaft/releases/20190521103931/node_modules/windshaft/lib/windshaft/renderers/renderer_factory.js:89:24
    at mmlBuilder.toXML (/home/ubuntu/www/node-windshaft/releases/20190521103931/node_modules/windshaft/lib/windshaft/renderers/mapnik/factory.js:226:20)
    at ChildProcess.done (/home/ubuntu/www/node-windshaft/releases/20190521103931/node_modules/grainstore/lib/grainstore/mml-builder/mml-builder.js:102:20)
    at Object.onceWrapper (events.js:277:13)
    at ChildProcess.emit (events.js:189:13)
    at emit (internal/child_process.js:820:12)
    at process._tickCallback (internal/process/next_tick.js:63:19)
```

These are the relevant source lines:

https://github.com/CartoDB/Windshaft/blob/a9b3b6492c4e01b16b12c807bf10f6d5d2661d32/lib/windshaft/renderers/blend/factory.js#L54

https://github.com/d3/d3-queue/blob/v1.1.0/src/queue.js#L34

(note there's a mismatch in the source and build lines, because of the build itself).

Not sure if it is a misuse or a bug in `queue-async`, but apparently it does not like the call to `done` when it is not running yet.

Mind that it is using a rather old version of `queue-async`.

At least a trace could be added to give visibility to the original error.

**Comment by dgaubert** [comment]
The main problem is `queue-async` is hiding the underlying error which comes from https://github.com/CartoDB/Windshaft/blob/master/lib/windshaft/renderers/mapnik/factory.js#L226, this error is thrown synchronously so I guess that's why `queue-async` considers the task is not running yet (the error happens before releasing the tick). I bet the original error comes from https://github.com/CartoDB/cartonik/blob/master/lib/renderer/raster-renderer.js#L19. However, we cannot know because `queue-async` is dumping the original error's the stack trace.

I'm going to research bit how hard will be get rid of `queue-async` for renderer's public interface.

**Comment by dgaubert** [comment]
I've been researching this issue and the right fix will be to remove `queue-async` as a dependency and use Promises as the standard to manage asynchronous code. Furthermore `queue-async` is a deprecated module, renamed `d3-queue`. But removing it properly means we need to change the signature of `.getRendender` and `.getTile` to return Promises, which requires tons of work and will be a project itself. We can wrap them from callback to promises but is going to be a hack and I don't want to add more entropy to this project.

So I'm going to upgrade `queue-async` to the latest release (1.2.1) before it was marked as deprecated, which comes with some fixes that might be related to this issue, and plan how to migrate from `callbacks` to `promises` in a project Issue.

**Comment by dgaubert** [comment]
I couldn't update `queue-async` to 1.2.1 because it had many side effects (multiple failing tests) so I decided to use `Promise.all` as a replacement. I needed to wrap some callback-ish method with promises. It isn't the best scenario as I already mentioned above but we'll end up to have `.getRenderer` & `.getTile`working with Promises natively so now we are one step closer...

**Comment by dgaubert** [comment]
Already fixed in #699
